pkg/typesystem/reachabilitygraph_test.go

```diff
@@ -17,6 +17,215 @@ import (
+		subjectType         string
+		relation            string
+		expectedPermissions []string
+	}{
+		{
+			"simple permission",
+			`definition user {}
+
+			definition document {
+				relation viewer: user
+				permission view = viewer
+			}`,
+
+			"document",
+			"viewer",
+			[]string{"document#view"},
+		},
+		{
+			"simple permission and user",
+			`definition user {}
+
+			definition document {
+				relation viewer: user
+				permission view = viewer
+			}`,
+
+			"user",
+			"...",
+			[]string{"document#view", "document#viewer"},
+		},
+		{
+			"multiple permissions using the same relation",
+			`definition user {}
+
+			definition document {
+				relation viewer: user
+				relation editor: user
+
+				permission edit = editor
+				permission view = viewer + editor
+			}`,
+
+			"document",
+			"editor",
+			[]string{"document#view", "document#edit"},
+		},
+		{
+			"multiple permissions using the same relation indirectly",
+			`definition user {}
+
+			definition document {
+				relation viewer: user
+				relation editor: user
+
+				permission edit = editor
+				permission view = viewer + edit
+			}`,
+
+			"document",
+			"editor",
+			[]string{"document#view", "document#edit"},
+		},
+		{
+			"permission referencing subject relation",
+			`definition user {}
+
+			definition group {
+				relation member: user
+			}
+
+			definition document {
+				relation viewer: user | group#member
+				permission view = viewer
+			}`,
+			"group",
+			"member",
+			[]string{"document#view", "document#viewer"},
+		},
+		{
+			"simple arrow",
+			`definition user {}
+
+			definition organization {
+				relation admin: user
+			}
+
+			definition document {
+				relation org: organization
+				permission view = org->admin
+			}`,
+			"organization",
+			"admin",
+			[]string{"document#view"},
+		},
+		{
+			"complex schema",
+			`definition user {}
+
+			definition organization {
+				relation admin: user
+				relation direct_member: user
+
+				permission member = direct_member + admin
+			}
+
+			definition group {
+				relation admin: user
+				relation direct_member: user | group#member
+				permission member = direct_member + admin
+			}
+
+			definition document {
+				relation viewer: user | group#member
+				relation owner: user
+				relation org: organization
+
+				permission view = viewer + owner + org->admin
+			}`,
+			"user",
+			"...",
+			[]string{
+				"document#viewer", "document#owner", "document#view",
+				"group#admin", "group#direct_member", "group#member",
+				"organization#member", "organization#admin",
+				"organization#direct_member",
+			},
+		},
+		{
+			"schema with different user types",
+			`definition user {}
+
+			definition platformuser {}
+
+			definition document {
+				relation viewer: user | platformuser
+				relation editor: platformuser
+
+				permission edit = editor
+				permission view = viewer + edit
+			}`,
+
+			"user",
+			"...",
+
+			[]string{"document#view", "document#viewer"},
+		},
+	}
```

Comment:
also add test where there is recursion in the schema. Also add to the other API tests

Reply:
We have one; a group refers to its own members